lib/cli_common/cli_common/phabricator.py

```diff
@@ -185,44 +187,113 @@ def load_parents(self, revision_phid):
+            buildTargetPHID=build_target_phid,
+            type=type,
+            unit=unit,
+            lint=lint,
+        )
+
+    def upload_coverage_results(self, object_phid, coverage_data):
         '''
+        Upload code coverage results to a Phabricator object.
+
+        `coverage_data` is an object in the format:
+        {
+            "this/is/a/path1": "UNCXUNCXUNCX",
+            "this/is/a/path2": "UUU",
+        }
+
+        The keys of the object are paths to the source files in the mozilla-central
+        repository.
+        The values are strings defining the coverage for each line of the source file
+        (one character per line), where:
+        - U means "not covered";
+        - N means "not executable";
+        - C means "covered";
+        - X means that no data is available about that line.
+        '''
+        # TODO: We are temporarily using arcanist.unit, but we should switch to something
+        # different after https://bugzilla.mozilla.org/show_bug.cgi?id=1487843 is resolved.
+        res = self.load_or_create_build_autotarget(object_phid, ['arcanist.unit'])
+        build_target_phid = res['arcanist.unit']
+
+        self.update_build_target(
+            build_target_phid,
+            'pass',
+            unit=[
+                {
+                    'name': 'Aggregate coverage information',
+                    'result': 'pass',
+                    'coverage': coverage_data,
+                }
+            ]
+        )
 
-        def flatten_params(params):
-            '''
-            Flatten nested objects and lists.
-            Phabricator requires query data in a application/x-www-form-urlencoded
-            format, so we need to flatten our params dictionary.
-            '''
-            assert isinstance(params, dict)
-            flat = {}
-            remaining = list(params.items())
-
-            # Run a depth-ish first search building the parameter name
-            # as we traverse the tree.
-            while remaining:
-                key, o = remaining.pop()
-                if isinstance(o, dict):
-                    gen = o.items()
-                elif isinstance(o, list):
-                    gen = enumerate(o)
-                else:
-                    flat[key] = o
-                    continue
-
-                remaining.extend(('{}[{}]'.format(key, k), v) for k, v in gen)
-
-            return flat
+    def upload_lint_results(self, object_phid, type, lint_data):
+        '''
+        Upload linting/static analysis results to a Phabricator object.
+
+        `type` is either "pass" if no errors were found, "fail" otherwise.
+
+        `lint_data` is an array of objects in the format:
+        {
+            "name": "Error name",
+            "code": "CODE1",
+            "severity": "error",
+            "path": "this/is/a/path.cpp",
+            "line": 42,
+            "char": 7,
+            "description": "A long description of the error."
+        }
```

Comment:
As `lint_data` is more complex than `coverage_data`, it could be interesting to use a `namedtuple` instead of a dict here. Define the namedtuple into this file, so we can use it as this:

```python
data = [
  LintData('name', 'CODE1', 'err', ...),
]
phabricator.upload_lint_results('PHID-XXX', 'pass', data)
```

Or you could validate that lint_data is correctly formatted ?

Reply:
With e5d4546b140bef4104e75560e61c5bd142373d29, I've made it an object to simplify JSON dumping in the right format.